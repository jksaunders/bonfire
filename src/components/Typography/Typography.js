import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { css } from '../../styling';
import { bonfireThemePropKey, withThemes } from '../../theming/ThemeContext';

export const TypographyContext = React.createContext({});

export const cssTypography = typographyPropKey => props => {
  const propsToCheck = typographyPropKey ? props[typographyPropKey] : props;
  return `
    -webkit-font-smoothing: antialiased;
    ${css('align', 'text-align')(propsToCheck)}
    ${css(
      ['bold', 'font-weight', 'bold'],
      ['weight', 'font-weight']
    )(propsToCheck)}
    ${css('color')(propsToCheck)}
    ${css('font', 'font-family')(propsToCheck)}
    ${css('italic', 'font-style', 'italic')(propsToCheck)}
    ${css('letterSpacing', 'letter-spacing')(propsToCheck)}
    ${css('lineHeight', 'line-height')(propsToCheck)}
    ${css('size', 'font-size')(propsToCheck)}
    ${css('transform', 'text-transform')(propsToCheck)}
  `;
};

const typographyPropTypes = {
  align: PropTypes.string,
  bold: PropTypes.bool,
  color: PropTypes.string,
  font: PropTypes.string,
  h1: PropTypes.bool,
  h2: PropTypes.bool,
  h3: PropTypes.bool,
  h4: PropTypes.bool,
  h5: PropTypes.bool,
  h6: PropTypes.bool,
  italic: PropTypes.bool,
  letterSpacing: PropTypes.string,
  lineHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  paragraph: PropTypes.bool,
  spaceAfter: PropTypes.bool,
  size: PropTypes.string,
  transform: PropTypes.string,
  weight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

const propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  variant: PropTypes.oneOfType([
    PropTypes.shape(typographyPropTypes),
    PropTypes.string,
  ]),
  ...typographyPropTypes,
};

const defaultProps = {
  /* eslint-disable react/default-props-match-prop-types */
  align: null,
  bold: null,
  color: null,
  font: null,
  h1: null,
  h2: null,
  h3: null,
  h4: null,
  h5: null,
  h6: null,
  italic: null,
  letterSpacing: null,
  lineHeight: null,
  paragraph: null,
  spaceAfter: null,
  size: null,
  transform: null,
  weight: null,
  /* eslint-enable react/default-props-match-prop-types */

  children: null,
  className: null,
  variant: null,
};

const StyledTypography = styled.span`
  margin: 0;
  ${css('spaceAfter', 'margin-block-start', '1em')}
  ${css('spaceAfter', 'margin-block-end', '1em')}
  ${css('spaceAfter', 'margin-inline-start', '0px')}
  ${css('spaceAfter', 'margin-inline-end', '0px')}
  ${cssTypography()}
`;

const getBaseElement = (variant, props) => {
  let baseElement = 'span';

  if ((variant && variant.h1) || props.h1) {
    baseElement = 'h1';
  } else if ((variant && variant.h2) || props.h2) {
    baseElement = 'h2';
  } else if ((variant && variant.h3) || props.h3) {
    baseElement = 'h3';
  } else if ((variant && variant.h4) || props.h4) {
    baseElement = 'h4';
  } else if ((variant && variant.h5) || props.h5) {
    baseElement = 'h5';
  } else if ((variant && variant.h6) || props.h6) {
    baseElement = 'h6';
  } else if ((variant && variant.paragraph) || props.paragraph) {
    baseElement = 'p';
  }

  return baseElement;
};

const getVariant = (variant, props) => {
  const result = {};
  const themeVariant =
    typeof variant === 'string' && props[bonfireThemePropKey] != null
      ? props[bonfireThemePropKey].getCurrentVariant('typography', variant)
      : null;

  Object.keys(typographyPropTypes).forEach(k => {
    if (props[k] != null) {
      result[k] = props[k];
    } else if (themeVariant != null) {
      result[k] = themeVariant[k];
    } else if (
      typeof variant === 'object' &&
      variant != null &&
      variant[k] != null
    ) {
      result[k] = variant[k];
    }
  });

  return result;
};

const Typography = ({ children, className, variant, ...props }) => (
  <StyledTypography
    className={className}
    {...getVariant(variant, props)}
    as={getBaseElement(variant, props)}
  >
    {children}
  </StyledTypography>
);

Typography.propTypes = propTypes;
Typography.defaultProps = defaultProps;
export default withThemes(Typography);
