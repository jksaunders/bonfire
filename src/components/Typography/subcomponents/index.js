export { default as H1 } from "./h1";
export { default as H2 } from "./h2";
export { default as H3 } from "./h3";
export { default as H4 } from "./h4";
export { default as H5 } from "./h5";
export { default as H6 } from "./h6";
export { default as Subtitle1 } from "./subtitle1";
export { default as Subtitle2 } from "./subtitle2";
export { default as Body1 } from "./body1";
export { default as Body2 } from "./body2";
export { default as Caption } from "./caption";
export { default as ButtonText } from "./button";
export { default as Overline } from "./overline";