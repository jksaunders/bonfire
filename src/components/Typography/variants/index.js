import * as ImportedMaterialVariants from './MaterialVariants';
import * as ImportedTrelloVariants from './TrelloVariants';

export const MaterialVariants = ImportedMaterialVariants;
export const TrelloVariants = ImportedTrelloVariants;