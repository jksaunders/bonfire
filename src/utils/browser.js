const goTo = (url) => {
  if (url && typeof window !== 'undefined') {
    window.location.href = url;
  }
};

export default {
  goTo
};