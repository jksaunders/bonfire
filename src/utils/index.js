export { default as Browser } from "./browser";
export * from "./styling";