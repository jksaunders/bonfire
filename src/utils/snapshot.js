/* eslint-disable import/no-extraneous-dependencies */
import { render as importedRender } from '@testing-library/react';
import 'jest-styled-components';

export const render = component => {
  const rendered = importedRender(component);
  return {
    ...rendered,
    component: rendered.container.firstChild,
  };
};

export const expectSnapshot = component => {
  expect(component).toMatchSnapshot();
};

export const expectExists = component => {
  expect(component).not.toEqual(null);
  expect(component).toBeInTheDocument();
};

export const expectStyle = (component, key, value) => {
  expect(component).toHaveStyleRule(key, value);
};
