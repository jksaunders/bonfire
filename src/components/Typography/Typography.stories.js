/* eslint-disable max-len */
import React from "react";
import { ThemeProvider } from "styled-components";
import { select, withKnobs } from "@storybook/addon-knobs";
import { ThemeConstants } from "../../theming";
import Typography, { MaterialVariants } from ".";

export default {
  title: "Style|Typography",
  component: Typography,
  decorators: [withKnobs]
};

export const AllStyles = () => {
  const themeMode = select("Theme Mode", Object.keys(ThemeConstants.mode.values), Object.keys(ThemeConstants.mode.values)[0]);
  const themeLayout = select("Theme Layout", Object.keys(ThemeConstants.layout.values), Object.keys(ThemeConstants.layout.values)[0]);
  const theme = { mode: themeMode, layout: themeLayout };

  return (
    <ThemeProvider theme={theme}>
      <div>
        <div>
          <Typography variant={MaterialVariants.H1}>h1</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.H2}>h2</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.H3}>h3</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.H4}>h4</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.H5}>h5</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.H6}>h6</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Subtitle1}>subtitle1</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Subtitle2}>subtitle2</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Body1}>body1</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Body2}>body2</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Caption}>caption</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.ButtonText}>button</Typography>
        </div>
        <div>
          <Typography variant={MaterialVariants.Overline}>overline</Typography>
        </div>
      </div>
    </ThemeProvider>
  );
};

export const SampleText = () => {
  const themeMode = select("Theme Mode", Object.keys(ThemeConstants.mode.values), Object.keys(ThemeConstants.mode.values)[0]);
  const themeLayout = select("Theme Layout", Object.keys(ThemeConstants.layout.values), Object.keys(ThemeConstants.layout.values)[0]);
  const theme = { mode: themeMode, layout: themeLayout };

  return (
    <ThemeProvider theme={theme}>
      <div>
        <div>
          <Typography variant={MaterialVariants.H1}>Welcome to bonfire</Typography>
          <Typography variant={MaterialVariants.H3}>This is its typography.</Typography>
          <Typography variant={MaterialVariants.Body1}>
            Typography is the art and technique of arranging type to make written language legible, readable, and appealing when displayed. The arrangement of type involves selecting typefaces, point sizes, line lengths, line-spacing (leading), and letter-spacing (tracking), and adjusting the space between pairs of letters (kerning). The term typography is also applied to the style, arrangement, and appearance of the letters, numbers, and symbols created by the process. Type design is a closely related craft, sometimes considered part of typography; most typographers do not design typefaces, and some type designers do not consider themselves typographers. Typography also may be used as a decorative device, unrelated to communication of information.
          </Typography>
          <Typography variant={MaterialVariants.Body1}>
            Typography is the work of typesetters (also known as compositors), typographers, graphic designers, art directors, manga artists, comic book artists, graffiti artists, and, now, anyone who arranges words, letters, numbers, and symbols for publication, display, or distribution, from clerical workers and newsletter writers to anyone self-publishing materials. Until the Digital Age, typography was a specialized occupation. Digitization opened up typography to new generations of previously unrelated designers and lay users. As the capability to create typography has become ubiquitous, the application of principles and best practices developed over generations of skilled workers and professionals has diminished. So at a time when scientific techniques can support the proven traditions (e.g., greater legibility with the use of serifs, upper and lower case, contrast, etc.) through understanding the limitations of human vision, typography as often encountered may fail to achieve its principal objective: effective communication.
          </Typography>
        </div>
      </div>
    </ThemeProvider>
  );
};