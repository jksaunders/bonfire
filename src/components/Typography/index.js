import Typography, {
  cssTypography as ImportedCssTypography,
  TypographyContext as ImportedTypographyContext,
} from './Typography';
import {
  MaterialVariants as ImportedMaterialVariants,
  TrelloVariants as ImportedTrelloVariants,
} from './variants';

export default Typography;
export const cssTypography = ImportedCssTypography;
export const TypographyContext = ImportedTypographyContext;
export const MaterialVariants = ImportedMaterialVariants;
export const TrelloVariants = ImportedTrelloVariants;
