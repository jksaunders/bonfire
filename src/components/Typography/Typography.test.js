import React from 'react';
import { expectExists, expectSnapshot, render } from '../../utils/snapshot';
import Typography from './Typography';

test('Typography renders', () => {
  const { component } = render(<Typography>Content</Typography>);
  expectExists(component);
  expectSnapshot(component);
});

describe('Props', () => {
  [
    { prop: 'align', propValue: 'center', key: 'text-align' },
    {
      prop: 'bold',
      propValue: true,
      key: 'font-weight',
      value: 'bold',
    },
    { prop: 'color', propValue: '#123456' },
    { prop: 'font', propValue: 'arial', key: 'font-family' },
    {
      prop: 'italic',
      propValue: true,
      key: 'font-style',
      value: 'italic',
    },
    { prop: 'letterSpacing', propValue: '1em', key: 'letter-spacing' },
    { prop: 'lineHeight', propValue: '1em', key: 'line-height' },
    { prop: 'size', propValue: '20px', key: 'font-size' },
    { prop: 'transform', propValue: 'uppercase', key: 'text-transform' },
    { prop: 'weight', propValue: '200', key: 'font-weight' },
    { prop: 'weight', propValue: 300, key: 'font-weight' },
  ].forEach(testCase => {
    test(`${testCase.prop}`, () => {
      const { component } = render(
        <Typography {...{ [testCase.prop]: testCase.propValue }}>
          Content
        </Typography>
      );
      expectExists(component);
      expectSnapshot(component);
      expect(component).toHaveStyleRule(
        testCase.key ? testCase.key : testCase.prop,
        testCase.value ? testCase.value : `${testCase.propValue}`
      );
    });
  });
});

describe('Using variant', () => {
  const variant = {
    size: '6rem',
    font: '"Roboto","Helvetica","Arial",sans-serif',
    weight: 300,
    lineHeight: 1,
    letterSpacing: '-0.01562em',
  };

  test('no overrides', () => {
    const { component } = render(
      <Typography variant={variant}>Content</Typography>
    );
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('line-height', '1');
  });

  test('with overrides', () => {
    const { component } = render(
      <Typography h1 lineHeight={4} variant={variant}>
        Content
      </Typography>
    );
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('line-height', '4');
  });
});
