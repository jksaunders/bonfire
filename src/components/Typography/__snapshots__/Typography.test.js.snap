// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Props align 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  text-align: center;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props bold 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: bold;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props color 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  color: #123456;
}

<span
  class="c0"
  color="#123456"
>
  Content
</span>
`;

exports[`Props font 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-family: arial;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props italic 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-style: italic;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props letterSpacing 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  -webkit-letter-spacing: 1em;
  -moz-letter-spacing: 1em;
  -ms-letter-spacing: 1em;
  letter-spacing: 1em;
}

<span
  class="c0"
  letter-spacing="1em"
>
  Content
</span>
`;

exports[`Props lineHeight 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  line-height: 1em;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props size 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-size: 20px;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props transform 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  text-transform: uppercase;
}

<span
  class="c0"
  transform="uppercase"
>
  Content
</span>
`;

exports[`Props underlined 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props weight 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: 200;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Props weight 2`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: 300;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Typography renders 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
}

<span
  class="c0"
>
  Content
</span>
`;

exports[`Using variant no overrides 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: 300;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
  line-height: 1;
  font-size: 6rem;
}

<span
  class="c0"
  letter-spacing="-0.01562em"
>
  Content
</span>
`;

exports[`Using variant with overrides 1`] = `
.c0 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: 300;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
  line-height: 4;
  font-size: 6rem;
}

<h1
  class="c0"
  letter-spacing="-0.01562em"
>
  Content
</h1>
`;
